plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,314 @@
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION
+import com.mapbox.maps.plugin.viewport.experimental.data.DefaultViewportTransitionOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.FollowingViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.OverviewViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportStateImpl
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransitionImpl
+import com.mapbox.maps.plugin.viewport.experimental.transition.ImmediateViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransition
+import java.lang.IllegalStateException
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ *
+ * The Viewport plugin allows to track objects on a map.
+ *
+ * It provides a structured approach to organizing camera management logic into states and transitions between them.
+ *
+ * at any given time, the viewport is either:
+ *  - idle (not updating the camera)
+ *  - in a state (camera is being managed by a ViewportState)
+ *  - transitioning (camera is being managed by a ViewportTransition)
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStatusObservers = CopyOnWriteArrayList<ViewportStatusObserver>()
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+  private var currentCancelable: Cancelable? = null
+  private val transitions = HashMap<Pair<ViewportState?, ViewportState?>, ViewportTransition>()
+  private lateinit var delegateProvider: MapDelegateProvider
+  private lateinit var gesturesPlugin: GesturesPlugin
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        currentCancelable?.cancel()
+        currentCancelable = null
+        updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION)
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      // no-ops
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {
+      // no-ops
+    }
+  }
+
+  /**
+   * Returns list of registered states.
+   */
+  override val states: Set<ViewportState>
+    get() = internalStates.toSet()
```

Comment:
Why is toSet conversion needed?

Reply:
This is a left over when the internal states is not a set, removed `toSet()`